doc/link_event_damping/Link-event-damping-HLD.md

```diff
@@ -0,0 +1,313 @@
+
+**Link flaps when link starts Down**
+
+| Time    | Event         | Action      | Advertised link state | Timer Expiry|
+|-------- |---------------|-------------|-----------------------|-------------|
+|6:00:00  |Initial State  |N/A          |Down                   |N/A          |
+|6:00:20  |Link Up        |Link up event damped, Timer started for hold_time_up |Down |6:00:50  |
+|6:00:25  |Link Down      |Timer cancelled  |Down               |N/A          |
+|6:01:00  |Link Up        |Link up event damped, Timer started for hold_time_up |Down |6:01:30  |
+|6:01:30  |Timer expires  |Link up sent |Up                     |N/A          |
+
+#####  1.7.1.3. <a name='Pros/cons'></a>Pros/cons of SyncD implementation
+
+**Pros**
+- Does not require any additional vendor support. Proposed implementation is not expected to make the vendor any changes to cater to the needs of the feature.
+- Can be generic support in SyncD for all possible vendor SAI layers. Implementation is not needed to be customized or tailored based on the vendor SAI layers.
+- Intercepts the port state change notifications before any other listeners get them.
+
+**Cons**
+- SyncD monitors and acknowledges the configuration. It does not send the config to the SAI layer.
+
+
+#### 1.7.2. <a name='VendorSAIImplementation'></a>Vendor SAI Implementation (Alternate Approach)
+
+In this approach, implementation of link event damping will be in the vendor SAI layer. Link event damping configuration will be pushed to SAI using the SAI APIs. SAI APIs for link event damping config is not available, so it needs to be added. When vendor SAI receives the config, it will apply the config on the requested port and start the damping logic. Vendor SAI will also maintain counters for undamped transitions that it can export.
+
+![SAI link event damping path](sai_link_event_damping_path.jpg)
+
+Vendor SAI layer will be responsible for damping both the physical port and netdev port’s oper state change notification.
+
+#####  1.7.2.1. <a name='Pros/cons'></a> Pros/cons of Vendor SAI implementation
+
+**Pros**
+- Pushes the damping down the stack.
+- Orch agent can treat the config as any other port config and send it to SAI.
+
+**Cons**
+- Requires SAI interface modifications.
+- Different vendor implementations may be different.
+- Any change in the behavior will require all vendors to make the correcponding changes.
+
+
+###  1.8. <a name='SAIAPI'></a>SAI API
+
+**SyncD Implementation**
+
+There are no changes expected in the SAI API.
+
+**Vendor SAI Implementation**
+
+To push the link event damping configuration, 2 new SAI port attributes needed:
+- SAI_PORT_ATTR_HOLD_TIME_UP
+- SAI_PORT_ATTR_HOLD_TIME_DOWN
+
+To get the total count of pre and post damped transitions on a port that reflect the actual count of link oper state change and count of link events that was advertized by SAI after link event damping, new counter attributes will be needed.
```

Comment:
Will counters be defined later ?

They can include:
- Number of received UP events
- Number of received DOWN events
- Number of ignored/filtered UP events
- Number of ignored/filtered DOWN events
- Number of timer cancellations
- Number of delayed events (when sent to apps on timer expiration)  - option since (# of recvd DOWN - # of ignored DOWN events) can be used instead